### ISSUE TITLE
Choose the type argument for the future returned from an invocation of an async function

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#33213]
When an async function is invoked, a fresh instance of `Future<T>` for some `T` is returned at some point, and there is no user-written code which creates that future so the choice of `T` is not made by a developer. The specification currently leaves it unspecified which value to use for `T`:

```
If f is marked async (9), then a fresh instance (10.6.1) o implementing
the built-in class Future is associated with the invocation. Then the body
of f is executed until it either suspends or completes, at which point o is
returned.
```
This issue serves to recommend that we specify this type argument explicitly.

To get started we may consider a couple of loose ideas: `Future<T> f() async {...}` should return an instance of `Future<T>`. `void f() async {...}` should return an instance of `Future<void>`. `FutureOr<T> f() async {...}` should probably return an instance of `Future<T>`. Etc.

**Comment by eernstg** [comment]
@lrhn, @leafpetersen do you agree that we should specify this? I suppose it will fit as an update of dartLangSpec.tex directly, with no intervening feature specification.

**Comment by leafpetersen** [comment]
Yes, we should specify this.  I think the right answer is that the returned value should be a `Future<flatten(returnType)>`.

**Comment by eernstg** [comment]
Right, that would be as follows (`declared return type --> dynamic type of returned Future`):
```
void --> Future<void>
dynamic --> Future<dynamic>
Object --> Future<Object>
Future<T> --> Future<T>, for all T
FutureOr<T> --> Future<T>, for all T
```
The first three may seem slightly accidental, but they also seem "natural".

**Comment by eernstg** [comment]
I've created a [CL](https://dart-review.googlesource.com/c/sdk/+/56487) updating dartLangSpec.tex to say this (including, of course, similar rules for `sync*` and `async*`).

**Comment by eernstg** [comment]
Closing: [Said CL](https://dart-review.googlesource.com/c/sdk/+/56487) has now been landed as commit 8dca855de113f42167a057971b1a948b56b9057c.
